### ISSUE TITLE
Config loading locked for a long time

### ISSUE THREAD

**Issue by QuentinFarizon** [magento/magento2#22504]
### Preconditions (*)
Magento : 2.2.8 EE
PHP: 7.1
Redis: 3.2.11

I have an issue with config loading being stalled by the new locking mechanism introduced in 2.2.8 by https://github.com/magento/magento2/commit/8136dd003b0e4052bd5bea2cad48bb03cc930dff
I understand the need for such a locking mechanism, and it would be normal that at some times it locks for a few milliseconds waiting for the new configuration to be inserted (the retry interval is at 100ms, so it should not lock for more than one or two times)
Furthermore, it should be on rare occasions (not on any page/ajax call)

But I observe in production that every calls have to wait for an important number of retries, thus delaying calls by a several hundreds of milliseconds or even seconds.

Please see two examples got using Blackfire (green paddings are calls to usleep in System::lockedLoadData), by calling GET https://afrimarket.bj/customer/section/load/ :

![Screenshot from 2019-04-25 16-43-07](https://user-images.githubusercontent.com/33634848/56800912-7ef6fe80-681c-11e9-8b69-9141068e7189.png)

![Screenshot from 2019-04-26 11-51-24](https://user-images.githubusercontent.com/33634848/56801180-1f4d2300-681d-11e9-92e7-ef3718f253b2.png)

I may not be easy to reproduce since you need a system on heavy load (I do not reproduce it on our staging for example) but I'm open to any idea / suggestion / insight that can help me with the resolution of this issue.  

### Steps to reproduce (*)
Do any ajax call

### Expected result (*)
Not more than one or two calls to usleep in System::lockedLoadData on a given endpoint/page 

### Actual result (*)
Several calls to usleep in System::lockedLoadData are done on each endpoint/page

Thanks !

**Comment by QuentinFarizon** [comment]
Please find other examples :

![Screenshot from 2019-04-26 14-41-57](https://user-images.githubusercontent.com/33634848/56815744-5aae1880-6842-11e9-94bc-81cd03a0803a.png)

![Screenshot from 2019-04-26 16-41-59](https://user-images.githubusercontent.com/33634848/56815754-600b6300-6842-11e9-8c24-37bf0ba07e36.png)

**Comment by QuentinFarizon** [comment]
I have tried calling flushall on redis to be sure to start with a clean state, with no luck

**Comment by mariuscris** [comment]
Hello there,

I want to let you know that we have the same issue with our customer shop. 

The shop uses:
- 2.3.1 Magento Commerce (EE)
- PHP 7.1.28
- Redis 3.2.11
- MySQL 5.6.36

Here is a screenshot of our new relic detailed issue:

![Screenshot 2019-06-05 at 10 11 12](https://user-images.githubusercontent.com/8698833/58940961-17b55e00-877b-11e9-9982-b826d1315bf1.png)

The strange part is that is not occurring all the time. 

Thank you.

**Comment by QuentinFarizon** [comment]
Hello @mariuscris 

I have mitigated the issue by patching a behaviour in Enterprise version, where each time a back-office agent restricted to a certain website does any action, it flushes the whole config cache (call to StoreManagerInterface::reinitStores), which needless to say is a very bad idea.

I still see the issue occurring, I suspect there is similar bad code flushing the whole config cache, but haven't been able to track what code was responsible.

**Comment by mariuscris** [comment]
Hello @QuentinFarizonAfrimarket,

Thank you for your response and helpful suggestion. 
I found a way to reproduce this effect in our customer shop. 

Every time when you press the "Save config" button from Stores -> Configuration admin page and during this save request you quickly reload separately a shop page (or someone else uses the shop during the request), than we had the above issue with loading time. 
Once the save request is done, the next page is loading normally. In our tests we used an admin user with Administrative role. Another mention regarding this test case is that we didn't need to modify a configuration and save it. We kept the same existing config and pressed save button. 

I didn't determined what particular code could cause this yet. This issue happened with Production or Developer mode set on, and all caches enabled.

**Comment by QuentinFarizon** [comment]
Hello @mariuscris 
Well what you're describing here is the normal scenario, and the reason they added the locking mechanism.
Before, when you tried to retrieve configurations that were not yet cached in redis, it was retrieved from the database, which could cause massive overload on the database.
Now, it waits for the config to be written in redis, and then retrieves it. I think it's far better this way.

Writing the config to redis should be quick enough (avoiding huge config values, like html snippets, certainly helps) and changing configuration in production is something that you won't do more than a few times a day at worst.

The issue we encountered and why I opened this ticket is that I see this config locking even when not changing/saving the configuration, and at heavy rate (configuration is written to redis several times a minute)

**Comment by hostep** [comment]
It looks like this PR is trying to resolve this issue I think, it sounds very related at least: https://github.com/magento/magento2/pull/22829
And https://github.com/magento/magento2/issues/22824 is probably a duplicate of this issue.

**Comment by QuentinFarizon** [comment]
Nice @hostep thanks, it should indeed be an improvement, so that all the requests do not get stuck writing.

But it won't solve (unless I'm mistaken) the fact that there shouldn't be so much config write, when no config is being changed by a human

**Comment by ilnytskyi** [comment]
We have similar or the same issue during order placement and customer section update
![Selection_298](https://user-images.githubusercontent.com/3192770/72246218-89e40000-35f2-11ea-8c59-dae8fc0d2d21.png)
another profile
![Selection_297](https://user-images.githubusercontent.com/3192770/72246226-8cdef080-35f2-11ea-82a4-c7c0012335c8.png)

**Comment by ilnytskyi** [comment]
Probably found additional way to reproduce the issue when looked through all request profiles in order flow
![Selection_296](https://user-images.githubusercontent.com/3192770/72247620-91f16f00-35f5-11ea-974a-e6c93be7181d.png)
The module https://amasty.com/custom-order-number-for-magento-2.html#changelog
Cleans the config cache in this url
`POST /rest/store_code/V1/guest-carts/Dx7aqUmYJyd9TvcjIro4ZSmDk5qDlY/payment-information`
Here is the highlighted timeline from PHP-SPX
![Selection_300](https://user-images.githubusercontent.com/3192770/72247578-75edcd80-35f5-11ea-87d2-3d6fde2102e4.png)
@AmastyLtd Could you verify this in your module ? I do not think we need to clean config cache after order placement

**Comment by ihor-sviziev** [comment]
@QuentinFarizon thank you for report!
I'm closing this issue as duplicate of https://github.com/magento/magento2/issues/22824, that was already confirmed.
